contracts/protocol/modules/v1/AaveV3LeverageModule.sol

```diff
@@ -0,0 +1,1147 @@
+        _addCollateralAssets(_setToken, _collateralAssets);
+        _addBorrowAssets(_setToken, _borrowAssets);
+    }
+
+    /**
+     * @dev MANAGER ONLY: Removes this module from the SetToken, via call by the SetToken. Any deposited collateral assets
+     * are disabled to be used as collateral on Aave. Aave Settings and manager enabled assets state is deleted.
+     * Note: Function will revert is there is any debt remaining on Aave
+     */
+    function removeModule() external override onlyValidAndInitializedSet(ISetToken(msg.sender)) {
+        ISetToken setToken = ISetToken(msg.sender);
+
+        // Sync Aave and SetToken positions prior to any removal action
+        sync(setToken);
+
+        address[] memory borrowAssets = enabledAssets[setToken].borrowAssets;
+        for(uint256 i = 0; i < borrowAssets.length; i++) {
+            IERC20 borrowAsset = IERC20(borrowAssets[i]);
+            require(underlyingToReserveTokens[borrowAsset].variableDebtToken.balanceOf(address(setToken)) == 0, "VDR");
+
+            delete borrowAssetEnabled[setToken][borrowAsset];
+        }
+
+        address[] memory collateralAssets = enabledAssets[setToken].collateralAssets;
+        for(uint256 i = 0; i < collateralAssets.length; i++) {
+            IERC20 collateralAsset = IERC20(collateralAssets[i]);
+            _updateUseReserveAsCollateral(setToken, collateralAsset, false);
+
+            delete collateralAssetEnabled[setToken][collateralAsset];
+        }
+
+        delete enabledAssets[setToken];
+
+        // Try if unregister exists on any of the modules
+        address[] memory modules = setToken.getModules();
+        for(uint256 i = 0; i < modules.length; i++) {
+            try IDebtIssuanceModule(modules[i]).unregisterFromIssuanceModule(setToken) {} catch {}
+        }
+    }
+
+    /**
+     * @dev MANAGER ONLY: Add registration of this module on the debt issuance module for the SetToken.
+     * Note: if the debt issuance module is not added to SetToken before this module is initialized, then this function
+     * needs to be called if the debt issuance module is later added and initialized to prevent state inconsistencies
+     * @param _setToken             Instance of the SetToken
+     * @param _debtIssuanceModule   Debt issuance module address to register
+     */
+    function registerToModule(ISetToken _setToken, IDebtIssuanceModule _debtIssuanceModule) external onlyManagerAndValidSet(_setToken) {
+        require(_setToken.isInitializedModule(address(_debtIssuanceModule)), "INI");
+
+        _debtIssuanceModule.registerToIssuanceModule(_setToken);
+    }
+
+    /**
+     * @dev MANAGER ONLY: Set EMode category
+     * @param _categoryId             CategoryId of the chosen EMode
+     */
+    function setEModeCategory(ISetToken _setToken, uint8 _categoryId) external onlyManagerAndValidSet(_setToken) {
+        _setToken.invokeSetUserEMode(IPool(lendingPoolAddressesProvider.getPool()), _categoryId);
+    }
```

Comment:
related here https://github.com/IndexCoop/index-protocol/pull/17#discussion_r1156535427

any comment on this new functionality that wasn't in the V2 Module?

what does it enable? how does it interact with other parts of the module? anything reviewers should key in on?

Reply:
See: https://github.com/IndexCoop/index-protocol/pull/17#discussion_r1158016497